model/src/main/scala/aqua/model/topology/Topology.scala

```diff
@@ -114,18 +80,90 @@ object Topology {
           nextOn.lastOption.filter(_ => parent == ParTag).map(_.peerId)
         )
 
-        if (prevOn.isEmpty && getThere.isEmpty) cfu :: Nil
-        else
-          (through(prevPath ++ loc.pathViaChain ++ getThere)
-            .append(cfu) ++ through(nextPath, reversed = true)).toList
+        if (prevOn.isEmpty && getThere.isEmpty) {
+          currentSpecifiedTree :: Nil
+        } else {
+          val wayBeforeFull = prevPathFlat ++ loc.pathViaChain ++ getThere ++ peerIdC
+
+          // filter optimized path by previous call peerId and current call peerId
+          // because on optimization they will stay on their's first and last positions
+          val optimizedWayBefore = optimizePath(wayBeforeFull).filter { vm =>
+            val containsTarget = peerIdC.contains(vm)
+            val containsPrevTargets = prevPath.find(_.lastOption.contains(vm)).isDefined
```

Comment:
- I don't understand this line.
- Can it be rewritten with `.exists`?
- Why only the `lastOption` counts?

Reply:
`lastOption` because we need to delete targets of `on` blocks that we added to `prevPath`